codegenerator/cli/npm/envio/src/bindings/ClickHouse.res

```diff
@@ -0,0 +1,387 @@
+
+  let entityFields =
+    entityConfig.table.fields
+    ->Belt.Array.keepMap(field => {
+      switch field {
+      | Field(field) => {
+          let fieldName = field->Table.getDbFieldName
+          Some(`\`${fieldName}\``)
+        }
+      | DerivedFrom(_) => None
+      }
+    })
+    ->Js.Array2.joinWith(", ")
+
+  `CREATE VIEW IF NOT EXISTS ${database}.\`${entityConfig.name}\` AS
+SELECT ${entityFields}
+FROM (
+  SELECT ${entityFields}, \`${EntityHistory.changeFieldName}\`
+  FROM ${database}.\`${historyTableName}\`
+  WHERE \`${EntityHistory.checkpointIdFieldName}\` <= (SELECT max(${checkpointIdField}) FROM ${database}.\`${checkpointsTableName}\`)
+  ORDER BY \`${EntityHistory.checkpointIdFieldName}\` DESC
+  LIMIT 1 BY \`${Table.idFieldName}\`
+)
+WHERE \`${EntityHistory.changeFieldName}\` = '${(EntityHistory.RowAction.SET :> string)}'`
+}
+
+// Initialize ClickHouse tables for entities
+let initialize = async (
+  client,
+  ~database: string,
+  ~entities: array<Internal.entityConfig>,
+  ~enums as _: array<Table.enumConfig<Table.enum>>,
+) => {
+  try {
+    await client->exec({query: `DROP DATABASE IF EXISTS ${database}`})
+    await client->exec({query: `CREATE DATABASE ${database}`})
+    await client->exec({query: `USE ${database}`})
+
+    await Promise.all(
+      entities->Belt.Array.map(entityConfig =>
+        client->exec({query: makeCreateHistoryTableQuery(~entityConfig, ~database)})
+      ),
+    )->Promise.ignoreValue
+    await client->exec({query: makeCreateCheckpointsTableQuery(~database)})
+
+    await Promise.all(
+      entities->Belt.Array.map(entityConfig =>
+        client->exec({query: makeCreateViewQuery(~entityConfig, ~database)})
+      ),
+    )->Promise.ignoreValue
+
+    Logging.trace("ClickHouse sink initialization completed successfully")
+  } catch {
+  | exn => {
+      Logging.errorWithExn(exn, "Failed to initialize ClickHouse sink")
+      Js.Exn.raiseError("ClickHouse initialization failed")
+    }
+  }
+}
+
+// Resume ClickHouse sink after reorg by deleting rows with checkpoint IDs higher than target
+let resume = async (client, ~database: string, ~checkpointId: float) => {
+  try {
+    // Try to use the database - will throw if it doesn't exist
+    try {
+      await client->exec({query: `USE ${database}`})
+    } catch {
+    | exn =>
+      Logging.errorWithExn(
+        exn,
+        `ClickHouse sink database "${database}" not found. Please run 'envio start -r' to reinitialize the indexer (it'll also drop Postgres database).`,
+      )
+      Js.Exn.raiseError("ClickHouse resume failed")
+    }
+
+    // Get all history tables
+    let tablesResult = await client->query({
+      query: `SHOW TABLES FROM ${database} LIKE '${EntityHistory.historyTablePrefix}%'`,
+    })
+    let tables: array<{"name": string}> = await tablesResult->json
+
+    // Delete rows with checkpoint IDs higher than the target for each history table
```

Comment:
Is there a state where the sink has lower checkpoint ids than the persistence?

Reply:
Currently no